File: 2022/python/day-01/day-01.py
#! /usr/bin/env python3

from typing import Iterable


def part1(input_file: str) -> int:
    """Determine largest number of calories carried by an Elf."""

    max_calories = 0
    cur_calories = 0
    with open(input_file) as f:
        for line in f:
            line = line.strip()

            if line:
                cur_calories += int(line)
            else:
                # empty line signifies that start of a new elf
                max_calories = max(cur_calories, max_calories)
                cur_calories = 0

    return max_calories


def part2(input_file: str) -> int:
    """Determine number of calories carried by top three Elves."""

    calories = []
    cur_calories = 0
    with open(input_file) as f:
        for line in f:
            line = line.strip()

            if line:
                cur_calories += int(line)
            else:
                # empty line signifies that start of a new elf
                calories.append(cur_calories)
                cur_calories = 0

    calories.append(cur_calories)

    return sum(sorted(calories, reverse=True)[0:3])


if __name__ == '__main__':
    assert part1('test.dat') == 24000
    print(f'Part 1: {part1("input.dat")}')

    assert part2('test.dat') == 45000
    print(f'Part 2: {part2("input.dat")}')

File: 2022/rust/day-01/src/day-01.rs
use itertools::Itertools;
use std::fs;

/// Determine maximum number of calories carried by an elf.
fn part1(data: &str) -> u32 {
    data.split("\n\n")
        .map(|elf_load| {
            elf_load
                .split_terminator('\n')
                .map(|item| item.parse::<u32>().unwrap())
                .sum()
        })
        .max()
        .unwrap()
}

/// Determine number of calories carried by 3 elves carrying the most calories.
fn part2(data: &str) -> u32 {
    data.split("\n\n")
        .map(|elf_load| {
            elf_load
                .split_terminator('\n')
                .map(|item| item.parse::<u32>().unwrap())
                .sum::<u32>()
        })
        .sorted_by(|a, b| b.cmp(a))
        .take(3)
        .sum()
}

fn main() {
    let data = fs::read_to_string("input.dat").unwrap();

    println!("Part 1: {}", part1(&data));
    println!("Part 2: {}", part2(&data));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let test_data = fs::read_to_string("test.dat").unwrap();
        let results = part1(&test_data);
        assert_eq!(results, 24000)
    }

    #[test]
    fn test_part2() {
        let test_data = fs::read_to_string("test.dat").unwrap();
        let results = part2(&test_data);
        assert_eq!(results, 45000)
    }
}
